vllm/entrypoints/openai/api_server.py

```diff
@@ -1840,23 +1840,51 @@ async def init_app_state(
     state.server_load_metrics = 0
 
 
-def create_server_socket(addr: tuple[str, int]) -> socket.socket:
-    family = socket.AF_INET
-    if is_valid_ipv6_address(addr[0]):
-        family = socket.AF_INET6
+def create_server_sockets(
+        addr: tuple[Optional[str], int]) -> list[socket.socket]:
+    """Create and bind server sockets for each resolved address."""
+
+    # This method is a subset of the asyncio.loop.create_server
+    # logic for name resolution for host binding.
+    sockets = []
+
+    # Avoid name resolution on already resolved IP addresses
+    infos: list[Union[tuple[socket.AddressFamily, socket.SocketKind, int, str,
+                            Union[tuple[str, int], tuple[str, int, int,
+                                                         int]]]]]
+    host, port = addr
+    if host and is_valid_ipv6_address(host):
+        infos = [(socket.AF_INET6, socket.SOCK_STREAM, socket.IPPROTO_TCP, "",
+                  (host, port))]
+    elif host and is_valid_ipv4_address(host):
+        infos = [(socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_TCP, "",
+                  (host, port))]
+    else:
+        infos = socket.getaddrinfo(host,
+                                   port,
+                                   type=socket.SOCK_STREAM,
+                                   flags=socket.AI_PASSIVE)
+
+    for family, socktype, proto, _, sa in infos:
+        sock = socket.socket(family, socktype, proto)
+        sockets.append(sock)
 
-    sock = socket.socket(family=family, type=socket.SOCK_STREAM)
-    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
-    sock.bind(addr)
+        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        # Since Linux 6.12.9, SO_REUSEPORT is not allowed
+        # on other address families than AF_INET/AF_INET6
+        if family in (socket.AF_INET, socket.AF_INET6):
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
 
-    return sock
+        logger.debug("binding to socket family=%d socktype=%d proto=%d: %s",
+                     family, socktype, proto, sa)
+        sock.bind(sa)
+    return sockets
```

Comment:
Hi @smarterclayton, I have a question about the error handling for edge cases here and I hope to get some clarifications about it. It looks like the new logic binds the sockets in a sequence. 
I was wondering what would happen if the first socket (ipv6) succeeds but the (ivpv4) socket fails to bind. 
In this case, from my reading the function would throw an exception and exits while the first socket (ipv6) is left intact without being cleaned up. 
Is my understanding correct or is there another part of the code that does this cleanup?